File: server/internal/cerebro/workflows/hook_types.go
package workflows

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

type HookEventType string

type HookDecision string

const (
	HookAllow   HookDecision = "allow"
	HookBlock   HookDecision = "block"
	HookModify  HookDecision = "modify"
	HookRequire HookDecision = "require"
)

type HookMode string

const (
	HookModeOff     HookMode = "off"
	HookModeDryRun  HookMode = "dry_run"
	HookModeEnforce HookMode = "enforce"
	HookModeManaged HookMode = "managed"
)

type HookConditionMode string

const (
	HookConditionAll HookConditionMode = "all"
	HookConditionAny HookConditionMode = "any"
)

type HookFailMode string

const (
	HookFailClosed HookFailMode = "closed"
	HookFailWarn   HookFailMode = "warn"
)

type HookLifecycleState string

const (
	HookLifecycleDraft         HookLifecycleState = "draft"
	HookLifecycleLive          HookLifecycleState = "live"
	HookLifecycleLiveWithDraft HookLifecycleState = "live_with_draft"
	HookLifecycleOff           HookLifecycleState = "off"
	HookLifecycleOffWithDraft  HookLifecycleState = "off_with_draft"
	HookLifecycleManaged       HookLifecycleState = "managed"
)

type HookLifecycle struct {
	State                HookLifecycleState `json:"state"`
	LivePolicyID         string             `json:"live_policy_id,omitempty"`
	LiveVersion          int                `json:"live_version,omitempty"`
	DraftID              string             `json:"draft_id,omitempty"`
	DraftSeriesID        string             `json:"draft_series_id,omitempty"`
	DraftRevision        int                `json:"draft_revision,omitempty"`
	LiveUnchangedByDraft bool               `json:"live_unchanged_by_draft"`
}

type HookScopeKind string

const (
	HookScopeWorkspace HookScopeKind = "workspace"
	HookScopeProject   HookScopeKind = "project"
	HookScopeWorkflow  HookScopeKind = "workflow"
	HookScopeAgent     HookScopeKind = "agent"
	HookScopeModel     HookScopeKind = "model"
	HookScopeIssue     HookScopeKind = "issue"
	HookScopeSession   HookScopeKind = "session"
)

type HookBinding struct {
	Kind     HookScopeKind `json:"kind"`
	ID       string        `json:"id,omitempty"`
	Priority int           `json:"priority,omitempty"`
}

type HookHandler struct {
	ID            string         `json:"id"`
	Decision      HookDecision   `json:"decision"`
	Requirement   string         `json:"requirement,omitempty"`
	Modifications map[string]any `json:"modifications,omitempty"`
	Actions       []HookAction   `json:"actions,omitempty"`
}

type HookAction struct {
	Type   string         `json:"type"`
	Config map[string]any `json:"config,omitempty"`
}

type HookActionStatus string

const (
	HookActionWouldRun HookActionStatus = "would_run"
	HookActionSuccess  HookActionStatus = "success"
	HookActionFailed   HookActionStatus = "failed"
	HookActionDenied   HookActionStatus = "denied"
)

type HookActionResult struct {
	Type           string           `json:"type"`
	Status         HookActionStatus `json:"status"`
	IdempotencyKey string           `json:"idempotency_key"`
	HandlerID      string           `json:"handler_id,omitempty"`
	ActionIndex    int              `json:"action_index"`
	Config         map[string]any   `json:"-"`
	Result         map[string]any   `json:"result,omitempty"`
	Error          string           `json:"error,omitempty"`
}

type HookPolicy struct {
	ID               string             `json:"id"`
	FamilyID         string             `json:"family_id"`
	DraftSeriesID    string             `json:"draft_series_id,omitempty"`
	Revision         int                `json:"revision,omitempty"`
	Version          int                `json:"version"`
	Name             string             `json:"name"`
	Description      string             `json:"description,omitempty"`
	ContractRule     string             `json:"contract_rule"`
	ContractSatisfy  string             `json:"contract_satisfy"`
	Mode             HookMode           `json:"mode"`
	FailMode         HookFailMode       `json:"fail_mode"`
	Events           []HookEventType    `json:"events"`
	Bindings         []HookBinding      `json:"bindings"`
	ConditionMode    HookConditionMode  `json:"condition_mode"`
	Conditions       []Condition        `json:"conditions,omitempty"`
	Handlers         []HookHandler      `json:"handlers"`
	CreatedByID      string             `json:"created_by_id,omitempty"`
	CreatedByType    string             `json:"created_by_type,omitempty"`
	UpdatedAt        time.Time          `json:"updated_at,omitempty"`
	LastRunAt        *time.Time         `json:"last_run_at,omitempty"`
	ObservedRuns     int                `json:"observed_run_count"`
	PassCount7D      int                `json:"pass_count_7d"`
	BlockCount7D     int                `json:"block_count_7d"`
	BaselineAt       *time.Time         `json:"baseline_at,omitempty"`
	CanPublish       bool               `json:"can_publish"`
	Lifecycle        HookLifecycle      `json:"lifecycle"`
	CompatibleEvents []HookJournalEvent `json:"compatible_events,omitempty"`
	// Live describes the version that is actually enforcing right now. It is
	// set only when the configuration above is a Draft, because in that case
	// every other field describes something that is NOT running yet.
	Live *HookPolicyLive `json:"live,omitempty"`
}

type HookPolicyLive struct {
	PolicyID    string       `json:"policy_id"`
	Version     int          `json:"version"`
	Name        string       `json:"name"`
	Decision    HookDecision `json:"decision"`
	Requirement string       `json:"requirement,omitempty"`
	FailMode    HookFailMode `json:"fail_mode"`
}

func (p *HookPolicy) UnmarshalJSON(data []byte) error {
	type hookPolicyAlias HookPolicy
	decoded := hookPolicyAlias{ConditionMode: HookConditionAll}
	if err := json.Unmarshal(data, &decoded); err != nil {
		return err
	}
	if decoded.ConditionMode == "" {
		decoded.ConditionMode = HookConditionAll
	}
	*p = HookPolicy(decoded)
	return nil
}

type HookActor struct {
	Type string `json:"type,omitempty"`
	ID   string `json:"id,omitempty"`
}

type HookEvent struct {
	EventID       string         `json:"event_id"`
	Type          HookEventType  `json:"event_type"`
	WorkspaceID   string         `json:"workspace_id,omitempty"`
	ProjectID     string         `json:"project_id,omitempty"`
	WorkflowID    string         `json:"workflow_id,omitempty"`
	AgentID       string         `json:"agent_id,omitempty"`
	Model         string         `json:"model,omitempty"`
	IssueID       string         `json:"issue_id,omitempty"`
	SessionID     string         `json:"session_id,omitempty"`
	Actor         HookActor      `json:"actor,omitempty"`
	Proposed      map[string]any `json:"proposed_action,omitempty"`
	BeforeState   map[string]any `json:"before_state,omitempty"`
	Context       map[string]any `json:"context,omitempty"`
	MutableFields []string       `json:"mutable_fields,omitempty"`
	Attempt       int            `json:"attempt,omitempty"`
	NoProgress    int            `json:"no_progress,omitempty"`
	HookDepth     int            `json:"hook_depth,omitempty"`
}

type HookMatch struct {
	PolicyID    string       `json:"policy_id"`
	PolicyName  string       `json:"policy_name,omitempty"`
	Version     int          `json:"version"`
	HandlerID   string       `json:"handler_id"`
	SourceScope HookBinding  `json:"source_scope"`
	Decision    HookDecision `json:"decision"`
	DryRun      bool         `json:"dry_run"`
	Idempotency string       `json:"idempotency_key"`
	MatchedAt   time.Time    `json:"matched_at"`
}

// HookRef names the enforcing policy behind a decision. Every user-visible
// message about a stopped action carries it, so nobody has to guess which of
// the workspace's hooks made the call.
type HookRef struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

func (r HookRef) Label() string {
	if r.Name != "" {
		return r.Name
	}
	return r.ID
}

type HookResult struct {
	RunID             string             `json:"run_id,omitempty"`
	Evaluated         bool               `json:"evaluated"`
	Decision          HookDecision       `json:"decision"`
	BlockedBy         *HookRef           `json:"blocked_by,omitempty"`
	WouldDecision     HookDecision       `json:"would_decision,omitempty"`
	Requirements      []string           `json:"requirements,omitempty"`
	Modifications     map[string]any     `json:"modifications,omitempty"`
	Matches           []HookMatch        `json:"matches,omitempty"`
	MatchedConditions []Condition        `json:"matched_conditions,omitempty"`
	TimedOut          bool               `json:"timed_out,omitempty"`
	Warning           string             `json:"warning,omitempty"`
	ActionResults     []HookActionResult `json:"action_results,omitempty"`
}

// blockingHook finds the enforcing policy that raised the decision to block or
// require. Dry-run matches are skipped: they never stop anything, so naming one
// would point at an innocent hook.
func (r HookResult) blockingHook() *HookRef {
	for _, match := range r.Matches {
		if !match.DryRun && (match.Decision == HookBlock || match.Decision == HookRequire) {
			return &HookRef{ID: match.PolicyID, Name: match.PolicyName}
		}
	}
	return nil
}

// ReasonWithHook is the one place a stopped action turns into a sentence a
// human can act on. It always names the hook when one is known, so "workflow
// hook blocked before.session.start" can never reach a user again.
func (r HookResult) ReasonWithHook(detail string) string {
	detail = strings.TrimSpace(detail)
	if r.BlockedBy == nil {
		return detail
	}
	if detail == "" {
		return fmt.Sprintf("Hook %q stopped this.", r.BlockedBy.Label())
	}
	return fmt.Sprintf("Hook %q stopped this: %s", r.BlockedBy.Label(), detail)
}

File: packages/cerebro-workflows/core/hook-types.ts
export type HookMode = "off" | "dry_run" | "enforce" | "managed";
export type HookFailMode = "open" | "closed" | "warn";
export type HookConditionMode = "all" | "any";
export type HookDecision = "allow" | "block" | "modify" | "require";
export type HookScopeKind = "workspace" | "project" | "workflow" | "agent" | "model" | "issue" | "session";
export type HookLifecycleState = "draft" | "live" | "live_with_draft" | "off" | "off_with_draft" | "managed";

export interface HookLifecycle {
  state: HookLifecycleState;
  live_policy_id?: string;
  live_version?: number;
  draft_id?: string;
  draft_series_id?: string;
  draft_revision?: number;
  live_unchanged_by_draft: boolean;
}

export type HookEventType =
  | "before.session.start"
  | "after.session.start"
  | "before.session.end"
  | "after.session.end"
  | "before.prompt.assemble"
  | "before.tool.call"
  | "after.tool.call"
  | "on.tool.failure"
  | "before.task.complete"
  | "before.agent.stop"
  | "before.subagent.start"
  | "after.subagent.stop"
  | "on.error"
  | "on.task.failure"
  | "before.wakeup.create"
  | "on.wakeup.fire_failure"
  | "before.issue.status_change"
  | "before.issue.assigned"
  | "before.message.send"
  | "after.workflow.step_completed";

export interface HookCondition {
  field: string;
  operator: string;
  value: string;
  conjunction?: "AND" | "OR";
}

export interface HookBinding {
  kind: HookScopeKind;
  value: string;
}

export interface HookActionDraft {
  type: string;
  label: string;
  config: Record<string, unknown>;
}

// Symbolic action targets, resolved per event by the server. Mirrors
// HookEventTargetLabels in server/internal/cerebro/workflows/hook_actions.go.
export const HOOK_EVENT_TARGET_LABELS: Record<string, string> = {
  "event.agent": "The agent that triggered this hook",
  "event.task": "The task that produced this event",
};

// What is actually enforcing right now. Present only when the hook body above
// is an unpublished Draft, so the list can stop describing a draft as live.
export interface HookLive {
  policy_id: string;
  version: number;
  name: string;
  decision: HookDecision;
  requirement?: string;
  fail_mode: HookFailMode;
}

export interface WorkflowHook {
  id?: string;
  family_id?: string;
  draft_series_id?: string;
  revision?: number;
  version: number;
  name: string;
  description: string;
  contract_rule: string;
  contract_satisfy: string;
  mode: HookMode;
  fail_mode: HookFailMode;
  events: HookEventType[];
  bindings: HookBinding[];
  condition_mode: HookConditionMode;
  conditions: HookCondition[];
  decision: HookDecision;
  requirement: string;
  actions: HookActionDraft[];
  baseline_run_count: number;
  pass_count_7d: number;
  block_count_7d: number;
  can_publish?: boolean;
  last_run_at?: string;
  lifecycle: HookLifecycle;
  compatible_events?: HookJournalEvent[];
  live?: HookLive;
}

// One row of the workspace-wide hook history.
export interface HookRunSummary {
  id: string;
  family_id: string;
  hook_name: string;
  policy_version: number;
  event_type: string;
  agent_id: string;
  issue_id: string;
  decision: HookDecision;
  would_decision?: HookDecision;
  enforced: boolean;
  requirements: string[];
  timed_out: boolean;
  latency_ms: number;
  created_at: string;
}

export interface HookJournalEvent {
  id: string;
  event_id: string;
  event_type: HookEventType;
  schema_version: number;
  occurred_at: string;
  expires_at: string;
}

export interface HookRun {
  id: string;
  created_at: string;
  source: string;
  policy_id: string;
  policy_version: number;
  source_scope: { kind: string; id: string };
  matched_conditions: string[];
  matched_steps: string[];
  decision: HookDecision;
  would_action: string;
  fail_mode: HookFailMode;
  remediation: string[];
  // A run recorded from a Dry run or a replayed test changed nothing. A live
  // run did. Calling every row a test was the old panel's worst lie.
  dry_run: boolean;
  side_effects: false;
  latency_ms: number;
  event?: Record<string, unknown>;
}

export const HOOK_EVENT_OPTIONS: ReadonlyArray<{ value: HookEventType; label: string; description: string; advanced?: boolean }> = [
  { value: "before.task.complete", label: "Before task completes", description: "Check work immediately before a task is marked complete." },
  { value: "on.task.failure", label: "When a task fails", description: "React after a task ends with an error." },
  { value: "before.wakeup.create", label: "Before wakeup is created", description: "Check a scheduled wakeup before it is saved." },
  { value: "on.wakeup.fire_failure", label: "When a wakeup fails", description: "React when a scheduled wakeup cannot start its work." },
  { value: "before.issue.status_change", label: "Before issue status changes", description: "Check an issue before its status is changed." },
  { value: "before.issue.assigned", label: "Before an issue is handed to an agent", description: "Check an issue before work is started for the assigned agent." },
  { value: "before.message.send", label: "Before message is sent", description: "Check an outgoing message before delivery." },
  { value: "after.workflow.step_completed", label: "After workflow step completes", description: "React after a workflow finishes one step." },
  { value: "before.tool.call", label: "Before tool call", description: "Check a tool request before it runs.", advanced: true },
  { value: "before.session.start", label: "Before session starts", description: "Check a session before it starts.", advanced: true },
  { value: "after.session.start", label: "After session starts", description: "React after a session starts.", advanced: true },
  { value: "before.session.end", label: "Before session ends", description: "Check a session before it ends.", advanced: true },
  { value: "after.session.end", label: "After session ends", description: "React after a session ends.", advanced: true },
  { value: "before.prompt.assemble", label: "Before prompt is assembled", description: "Check context before the final prompt is assembled.", advanced: true },
  { value: "after.tool.call", label: "After tool call", description: "React after a tool returns successfully.", advanced: true },
  { value: "on.tool.failure", label: "When a tool fails", description: "React when a tool returns an error.", advanced: true },
  { value: "before.agent.stop", label: "Before agent stops", description: "Check an agent immediately before it stops.", advanced: true },
  { value: "before.subagent.start", label: "Before subagent starts", description: "Check delegated work before it starts.", advanced: true },
  { value: "after.subagent.stop", label: "After subagent stops", description: "React after delegated work stops.", advanced: true },
  { value: "on.error", label: "When an error occurs", description: "React to an unhandled workflow error.", advanced: true },
];

export function createHookDraft(): WorkflowHook {
  return {
    version: 1,
    name: "",
    description: "",
    contract_rule: "",
    contract_satisfy: "",
    mode: "dry_run",
    fail_mode: "warn",
    events: [],
    bindings: [],
    condition_mode: "all",
    conditions: [],
    decision: "allow",
    requirement: "",
    actions: [],
    baseline_run_count: 0,
    pass_count_7d: 0,
    block_count_7d: 0,
    can_publish: false,
    lifecycle: { state: "draft", live_unchanged_by_draft: false },
    compatible_events: [],
  };
}
